src/EditorFeatures/Core/Implementation/Classification/SyntacticClassificationTaggerProvider.TagComputer.cs

```diff
@@ -188,46 +198,80 @@ private void EnqueueProcessSnapshot(Document newDocument)
+                        // example, in C# and VB we will parse the file so that are called from tagger from UI thread,
+                        // we have the root of the tree ready to go.
+                        currentCachedData = await service.GetDataToCacheAsync(currentDocument, cancellationToken).ConfigureAwait(false);
+
+                        // Query the service to determine waht span of the document actually changed and should be
+                        // reclassified in the host editor.
+                        changedSpan = await GetChangedSpanAsync(currentDocument, currentSnapshot, service, cancellationToken).ConfigureAwait(false);
+                    }
                 }
 
+                // Once we're past this point, we're mutating our internal state so we cannot cancel past that.
+                cancellationToken = CancellationToken.None;
+
                 lock (_gate)
                 {
-                    _lastProcessedSnapshot = snapshot;
-                    _lastProcessedDocument = document;
+                    _lastProcessedSnapshot = currentSnapshot;
+                    _lastProcessedDocument = currentDocument;
+                    _lastProcessedCachedData = currentCachedData;
                 }
 
                 _reportChangeCancellationSource = new CancellationTokenSource();
                 _notificationService.RegisterNotification(() =>
                     {
                         _workQueue.AssertIsForeground();
-                        ReportChangedSpan(snapshot.GetFullSpan());
+                        ReportChangedSpan(changedSpan);
                     },
                     ReportChangeDelayInMilliseconds,
                     _listener.BeginAsyncOperation("ReportEntireFileChanged"),
                     _reportChangeCancellationSource.Token);
             }
 
+            private async Task<SnapshotSpan> GetChangedSpanAsync(
+                Document currentDocument, ITextSnapshot currentSnapshot,
+                IClassificationService service, CancellationToken cancellationToken)
+            {
+                // We don't need to grab _lastProcessedDocument in a lock.  We don't care which version of the previous
+                // doc we grab, just that we grab some prior version.  This is only used 
+                var previousDocument = _lastProcessedDocument;
+                if (previousDocument != null)
+                {
+                    var changeRange = await service.ComputeSyntacticChangeRangeAsync(
+                        previousDocument, currentDocument, _diffTimeout, cancellationToken).ConfigureAwait(false);
+                    if (changeRange != null)
+                        return currentSnapshot.GetSpan(changeRange.Value.Span.Start, changeRange.Value.NewLength);
+                }
+
+                // Couldn't compute a narrower range.  Just the mark the entire file as changed.
+                return currentSnapshot.GetFullSpan();
```

Comment:
For future people, what is the advice on using `GetChangedSpanAsync` over `currentSnapshot.GetFullSpan` in this class?

Reply:
Effectively (and i will doc this) no one should use GetFullSpan except for thsi code.